src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,244 @@
+                if (!ShouldAnalyze(
+                    (IInvocationOperation)context.Operation,
+                    cancellationTokenType,
+                    out string? cancellationTokenParameterName,
+                    out string? methodName))
+                {
+                    return;
+                }
+
+                context.ReportDiagnostic(context.Operation.CreateDiagnostic(ForwardCancellationTokenToAsyncMethodsRule, cancellationTokenParameterName!, methodName!));
+            },
+            OperationKind.Invocation);
+        }
+
+        private static bool ShouldAnalyze(
+            IInvocationOperation invocation,
+            INamedTypeSymbol cancellationTokenType,
+            out string? cancellationTokenParameterName,
+            out string? methodName)
+        {
+            IMethodSymbol method = invocation.TargetMethod;
+
+            cancellationTokenParameterName = null;
+            methodName = method.Name;
+
+            // Check if the invocation has an optional implicit ct or an overload that takes one ct
+            if (!InvocationIgnoresOptionalCancellationToken(cancellationTokenType, invocation, method) &&
+                !MethodHasCancellationTokenOverload(cancellationTokenType, method))
+            {
+                return false;
+            }
+
+            // Find the ancestor method that contains this invocation
+            if (!TryGetAncestorDeclaration(invocation, out IMethodSymbol? methodDeclaration))
+            {
+                return false;
+            }
+
+            // Check if the ancestor method has a ct that we can pass to the invocation
+            if (!VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(cancellationTokenType, methodDeclaration!, out cancellationTokenParameterName))
+            {
+                return false;
+            }
+
+            return cancellationTokenParameterName != null && methodName != null;
+        }
+
+        // Looks for an ancestor that could be a method or function declaration.
+        private static bool TryGetAncestorDeclaration(IInvocationOperation invocation, out IMethodSymbol? declaration)
+        {
+            declaration = null;
+
+            IOperation currentOperation = invocation.Parent;
+            while (currentOperation != null)
+            {
+                if (currentOperation.Kind == OperationKind.AnonymousFunction && currentOperation is IAnonymousFunctionOperation anonymousFunction)
+                {
+                    declaration = anonymousFunction.Symbol;
+                    break;
+                }
+                else if (currentOperation.Kind == OperationKind.LocalFunction && currentOperation is ILocalFunctionOperation localFunction)
+                {
+                    declaration = localFunction.Symbol;
+                    break;
+                }
+                else if (currentOperation.Kind == OperationKind.MethodBody && currentOperation is IMethodBodyOperation methodBody)
```

Comment:
You can remove this `else if` part completely and just return `context.ContainingSymbol` outside the while loop.

Reply:
Do you mind explaining this suggestion? Why would it work?

I suspect it's because we can expect all invocation operations to have a `OperationKind.Block` as their `ContainingSymbol`.
But can we expect all containing symbols to _always_ be `OperationKind.MethodBody`?